<a href="https://colab.research.google.com/github/dhanushpachabhatla/my-machine-learning-notes/blob/main/New_york_taxi_fair_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NYC TAXI FARE PREDICTION

# Steps to follow for this project
1. Download of Dataset
2. Explore & analyze the dataset
3. Prepare the dataset for ML training
4. Train hardcoded & Baseline models
5. Make Predictions and submit to kaggle
6. Perform feature engineering
7. Train & evaluate diferrent models
8. Tune hyperparameters for the best models
9. Train on GPU with entire dataset
10. Document & publish the project online

# 1. Download the Dataset

Steps  :

*   Install required lib
*   Download data from kaggle
*   view dataset files
*   Load training set with Pandas
*   Load Test set with Pandas

In [ ]:
!pip install jovian opendatasets pandas numpy scikit-learn xgboost --quiet

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 68.6/68.6 kB 2.0 MB/s eta 0:00:00


# Download the dataset form kaggle

We will use the opendatasets library to download the dataset. You will need to update your Kaggle API key to collab

In [ ]:
import opendatasets as od

dataset_url = 'https://www.kaggle.com/c/new-york-city-taxi-fare-prediction'  # Updated URL
od.download(dataset_url)

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username: dhanushpachabhatla
Your Kaggle Key: ··········
Extracting archive ./new-york-city-taxi-fare-prediction/new-york-city-taxi-fare-prediction.zip to ./new-york-city-taxi-fare-prediction


In [ ]:
data_dir = './new-york-city-taxi-fare-prediction'

# View dataset files

!ls -lh - info about whole dataseta and files it contain

In [ ]:
!ls -lh {data_dir}

total 5.4G
-rw-r--r-- 1 root root  486 Mar 25 16:54 GCP-Coupons-Instructions.rtf
-rw-r--r-- 1 root root 336K Mar 25 16:54 sample_submission.csv
-rw-r--r-- 1 root root 960K Mar 25 16:54 test.csv
-rw-r--r-- 1 root root 5.4G Mar 25 16:55 train.csv


!wc -l gives you total number of rows of your datset

In [ ]:
!wc -l {data_dir}/train.csv

55423856 ./new-york-city-taxi-fare-prediction/train.csv


In [ ]:
!wc -l {data_dir}/test.csv

9914 ./new-york-city-taxi-fare-prediction/test.csv


In [ ]:
!wc -l {data_dir}/sample_submission.csv

9915 ./new-york-city-taxi-fare-prediction/sample_submission.csv


!head - you will get first 10 rows of the data

In [ ]:
!head {data_dir}/train.csv

key,fare_amount,pickup_datetime,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,passenger_count
2009-06-15 17:26:21.0000001,4.5,2009-06-15 17:26:21 UTC,-73.844311,40.721319,-73.84161,40.712278,1
2010-01-05 16:52:16.0000002,16.9,2010-01-05 16:52:16 UTC,-74.016048,40.711303,-73.979268,40.782004,1
2011-08-18 00:35:00.00000049,5.7,2011-08-18 00:35:00 UTC,-73.982738,40.76127,-73.991242,40.750562,2
2012-04-21 04:30:42.0000001,7.7,2012-04-21 04:30:42 UTC,-73.98713,40.733143,-73.991567,40.758092,1
2010-03-09 07:51:00.000000135,5.3,2010-03-09 07:51:00 UTC,-73.968095,40.768008,-73.956655,40.783762,1
2011-01-06 09:50:45.0000002,12.1,2011-01-06 09:50:45 UTC,-74.000964,40.73163,-73.972892,40.758233,1
2012-11-20 20:35:00.0000001,7.5,2012-11-20 20:35:00 UTC,-73.980002,40.751662,-73.973802,40.764842,1
2012-01-04 17:22:00.00000081,16.5,2012-01-04 17:22:00 UTC,-73.9513,40.774138,-73.990095,40.751048,1
2012-12-03 13:10:00.000000125,9,2012-12-03 13:10:00 UTC,-74.006462,40.726713,-73.99

In [ ]:
!head {data_dir}/test.csv

key,pickup_datetime,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,passenger_count
2015-01-27 13:08:24.0000002,2015-01-27 13:08:24 UTC,-73.973320007324219,40.7638053894043,-73.981430053710938,40.74383544921875,1
2015-01-27 13:08:24.0000003,2015-01-27 13:08:24 UTC,-73.986862182617188,40.719383239746094,-73.998886108398438,40.739200592041016,1
2011-10-08 11:53:44.0000002,2011-10-08 11:53:44 UTC,-73.982524,40.75126,-73.979654,40.746139,1
2012-12-01 21:12:12.0000002,2012-12-01 21:12:12 UTC,-73.98116,40.767807,-73.990448,40.751635,1
2012-12-01 21:12:12.0000003,2012-12-01 21:12:12 UTC,-73.966046,40.789775,-73.988565,40.744427,1
2012-12-01 21:12:12.0000005,2012-12-01 21:12:12 UTC,-73.960983,40.765547,-73.979177,40.740053,1
2011-10-06 12:10:20.0000001,2011-10-06 12:10:20 UTC,-73.949013,40.773204,-73.959622,40.770893,1
2011-10-06 12:10:20.0000003,2011-10-06 12:10:20 UTC,-73.777282,40.646636,-73.985083,40.759368,1
2011-10-06 12:10:20.0000002,2011-10-06 12:10:20 UTC,-74.01409

In [ ]:
!head {data_dir}/sample_submission.csv

key,fare_amount
2015-01-27 13:08:24.0000002,11.35
2015-01-27 13:08:24.0000003,11.35
2011-10-08 11:53:44.0000002,11.35
2012-12-01 21:12:12.0000002,11.35
2012-12-01 21:12:12.0000003,11.35
2012-12-01 21:12:12.0000005,11.35
2011-10-06 12:10:20.0000001,11.35
2011-10-06 12:10:20.0000003,11.35
2011-10-06 12:10:20.0000002,11.35


Observations :

*   This is a supervides regression problem
*   Training data is of 5.5 gb
*   Training data contains 5.5 million rows
*   Testing data have rows < 10000
*   Training set have columns  :
key,fare_amount,pickup_datetime,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,passenger_count

*   Testing have same except fare_amount
*   The submission file should only contain the key and fair_amount for each test sample

# Loading Training Set

Tip - When working with large datasets, always start with a sample tp experiment & iterate faster
*    Loading entire dataset into Pandas is gonig to be slow, so we can use the following optimizations :

*    ignore key column
*    parse pickup datetime while loading data
*    specify datatypes for other colums
*    work with 1% sample here - 500k rows


In [ ]:
import pandas as pd

In [ ]:
selected_cols = 'key,fare_amount,pickup_datetime,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,passenger_count'.split(',')

In [ ]:
selected_cols

['key',
 'fare_amount',
 'pickup_datetime',
 'pickup_longitude',
 'pickup_latitude',
 'dropoff_longitude',
 'dropoff_latitude',
 'passenger_count']

In [ ]:
sample_fraction = 0.01

In [ ]:
import random


In [ ]:
random.random()

0.4558721957538421

In [ ]:
dtypes = {
 'fare_amount' : 'float32',
 'pickup_longitude' : 'float32',
 'pickup_latitude' : 'float32',
 'dropoff_longitude' : 'float32',
 'dropoff_latitude' : 'float32',
 'passenger_count' : 'uint8'
}

def skip_rows(row_idx):
  if row_idx  == 0:
    return False
  return random.random() > sample_fraction

random.seed(42)

df = pd.read_csv(data_dir + '/train.csv',
                 usecols = selected_cols,
                 parse_dates=['pickup_datetime'],
                 dtype=dtypes,
                 skiprows=skip_rows)

In [ ]:
test_df = pd.read_csv(data_dir+ '/test.csv',
                 dtype=dtypes,
                 parse_dates=['pickup_datetime'])

In [ ]:
df

key  fare_amount           pickup_datetime  \
0         2014-12-06 20:36:22.0000008          4.0 2014-12-06 20:36:22+00:00   
1        2013-01-17 17:22:00.00000043          8.0 2013-01-17 17:22:00+00:00   
2       2011-06-15 18:07:00.000000186          8.9 2011-06-15 18:07:00+00:00   
3        2009-12-14 12:33:00.00000075          6.9 2009-12-14 12:33:00+00:00   
4         2013-11-06 11:26:54.0000002          7.0 2013-11-06 11:26:54+00:00   
...                               ...          ...                       ...   
552445    2014-02-06 23:59:45.0000001         45.0 2014-02-06 23:59:45+00:00   
552446    2015-01-05 15:29:08.0000002         22.5 2015-01-05 15:29:08+00:00   
552447   2013-02-17 22:27:00.00000092          4.5 2013-02-17 22:27:00+00:00   
552448    2013-01-27 12:41:00.0000009         14.5 2013-01-27 12:41:00+00:00   
552449   2014-10-18 07:51:00.00000060          6.0 2014-10-18 07:51:00+00:00   

        pickup_longitude  pickup_latitude  dropoff_longitude  \
0             -73.979813        40.751904         -73.979446   
1               0.000000         0.000000           0.000000   
2             -73.996330        40.753223         -73.978897   
3             -73.982430        40.745747         -73.982430   
4             -73.959061        40.781059         -73.962059   
...                  ...              ...                ...   
552445        -73.973587        40.747669         -73.999916   
552446        -73.935928        40.799656         -73.985710   
552447        -73.992531        40.748619         -73.998436   
552448        -74.012115        40.706635         -73.988724   
552449        -73.997681        40.724380         -73.994148   

        dropoff_latitude  passenger_count  
0              40.755482                1  
1               0.000000                2  
2              40.766964                3  
3              40.745747                1  
4              40.768604                1  
...                  ...              ...  
552445         40.602894                1  
552446         40.726952                2  
552447         40.740143                1  
552448         40.756218                1  
552449         40.717796                1  

[552450 rows x 8 columns]

In [ ]:
test_df

key          pickup_datetime  pickup_longitude  \
0     2015-01-27 13:08:24.0000002  2015-01-27 13:08:24 UTC        -73.973320   
1     2015-01-27 13:08:24.0000003  2015-01-27 13:08:24 UTC        -73.986862   
2     2011-10-08 11:53:44.0000002  2011-10-08 11:53:44 UTC        -73.982521   
3     2012-12-01 21:12:12.0000002  2012-12-01 21:12:12 UTC        -73.981163   
4     2012-12-01 21:12:12.0000003  2012-12-01 21:12:12 UTC        -73.966049   
...                           ...                      ...               ...   
9909  2015-05-10 12:37:51.0000002  2015-05-10 12:37:51 UTC        -73.968124   
9910  2015-01-12 17:05:51.0000001  2015-01-12 17:05:51 UTC        -73.945511   
9911  2015-04-19 20:44:15.0000001  2015-04-19 20:44:15 UTC        -73.991600   
9912  2015-01-31 01:05:19.0000005  2015-01-31 01:05:19 UTC        -73.985573   
9913  2015-01-18 14:06:23.0000006  2015-01-18 14:06:23 UTC        -73.988022   

      pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  
0           40.763805         -73.981430         40.743835                1  
1           40.719383         -73.998886         40.739201                1  
2           40.751259         -73.979652         40.746140                1  
3           40.767807         -73.990448         40.751637                1  
4           40.789776         -73.988564         40.744427                1  
...               ...                ...               ...              ...  
9909        40.796997         -73.955643         40.780388                6  
9910        40.803600         -73.960213         40.776371                6  
9911        40.726608         -73.789742         40.647011                6  
9912        40.735432         -73.939178         40.801731                6  
9913        40.754070         -74.000282         40.759220                6  

[9914 rows x 7 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 552450 entries, 0 to 552449
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype              
---  ------             --------------   -----              
 0   key                552450 non-null  object             
 1   fare_amount        552450 non-null  float32            
 2   pickup_datetime    552450 non-null  datetime64[ns, UTC]
 3   pickup_longitude   552450 non-null  float32            
 4   pickup_latitude    552450 non-null  float32            
 5   dropoff_longitude  552450 non-null  float32            
 6   dropoff_latitude   552450 non-null  float32            
 7   passenger_count    552450 non-null  uint8              
dtypes: datetime64[ns, UTC](1), float32(5), object(1), uint8(1)
memory usage: 19.5+ MB


In [ ]:
df.describe()

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  552450.000000     552450.000000    552450.000000      552450.000000   
mean       11.354059        -72.497063        39.910500         -72.504326   
std         9.810809         11.622035         8.041162          12.065184   
min       -52.000000      -1183.362793     -3084.490234       -3356.729736   
25%         6.000000        -73.992020        40.734875         -73.991425   
50%         8.500000        -73.981819        40.752621         -73.980179   
75%        12.500000        -73.967155        40.767036         -73.963737   
max       499.000000       2420.209473       404.983337        2467.752686   

       dropoff_latitude  passenger_count  
count     552450.000000    552450.000000  
mean          39.934265         1.684983  
std            9.226158         1.337664  
min        -2073.150635         0.000000  
25%           40.733990         1.000000  
50%           40.753101         1.000000  
75%           40.768059         2.000000  
max         3351.403076       208.000000

In [ ]:
df['pickup_datetime'].min(), df['pickup_datetime'].max()

(Timestamp('2009-01-01 00:11:46+0000', tz='UTC'),
 Timestamp('2015-06-30 23:59:54+0000', tz='UTC'))

Observations of training data :    
* No missing data
* some wrong info as laltitute & longitude are negative so need some data cleaning
* passenger count max is 208 which is not possibe - again some data cleaning
* date from jan 2009 to june 2015

# Test Set

In [ ]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9914 entries, 0 to 9913
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   key                9914 non-null   object 
 1   pickup_datetime    9914 non-null   object 
 2   pickup_longitude   9914 non-null   float32
 3   pickup_latitude    9914 non-null   float32
 4   dropoff_longitude  9914 non-null   float32
 5   dropoff_latitude   9914 non-null   float32
 6   passenger_count    9914 non-null   uint8  
dtypes: float32(4), object(2), uint8(1)
memory usage: 319.6+ KB


In [ ]:
test_df.describe()

pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
count       9914.000000      9914.000000        9914.000000       9914.000000   
mean         -73.974716        40.751041         -73.973656         40.751740   
std            0.042799         0.033542           0.039093          0.035436   
min          -74.252190        40.573143         -74.263245         40.568974   
25%          -73.992500        40.736125         -73.991249         40.735253   
50%          -73.982327        40.753052         -73.980015         40.754065   
75%          -73.968012        40.767113         -73.964062         40.768757   
max          -72.986534        41.709557         -72.990967         41.696682   

       passenger_count  
count      9914.000000  
mean          1.671273  
std           1.278747  
min           1.000000  
25%           1.000000  
50%           1.000000  
75%           2.000000  
max           6.000000

Ask and answer questions about the dataset:    
Ex :
* what is busiest day of week?
* what is besiest time of day?
* in which month are fares high?
* Which pickup locations have the highest fares?
* which drop locations have the highest fares?
* what is the average distance?

Perfrom EDA on your dataset

# Prepare dataset for Training

Since test set also have same range of dates of the training set, we will split the 20% data for validation randomly


* your val test should be simimlar to your test set as it will show you the real type of accuracty

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
train_df, val_df  = train_test_split(df, test_size=0.2, random_state=42)

In [ ]:
len(train_df), len(val_df)

(441960, 110490)

# Remove/Fill Missing Values - Imputation

There are no missing values in our case but if there were we can simply drop the rows with missinig values instead of trying to fill them

In [ ]:
train_df = train_df.dropna()
val_df = val_df.dropna()

# Extract Inputs and Outputs

Training


In [ ]:
train_df.columns

Index(['key', 'fare_amount', 'pickup_datetime', 'pickup_longitude',
       'pickup_latitude', 'dropoff_longitude', 'dropoff_latitude',
       'passenger_count'],
      dtype='object')

In [ ]:
input_cols = ['key', 'pickup_datetime', 'pickup_longitude',
       'pickup_latitude', 'dropoff_longitude', 'dropoff_latitude',
       'passenger_count']

In [ ]:
target_col = 'fare_amount'

In [ ]:
train_inputs = train_df[input_cols]
train_target = train_df[target_col]

In [ ]:
val_inputs = val_df[input_cols]
val_target = val_df[target_col]

In [ ]:
train_inputs

key           pickup_datetime  \
353352    2015-04-12 03:40:38.0000004 2015-04-12 03:40:38+00:00   
360070  2011-01-26 19:21:00.000000153 2011-01-26 19:21:00+00:00   
372609    2012-10-03 10:40:17.0000003 2012-10-03 10:40:17+00:00   
550895    2012-03-14 13:44:27.0000004 2012-03-14 13:44:27+00:00   
444151   2012-02-05 15:33:00.00000084 2012-02-05 15:33:00+00:00   
...                               ...                       ...   
110268   2009-09-06 16:12:00.00000051 2009-09-06 16:12:00+00:00   
259178    2009-04-12 09:58:56.0000001 2009-04-12 09:58:56+00:00   
365838  2012-07-12 19:30:00.000000266 2012-07-12 19:30:00+00:00   
131932   2011-02-17 18:33:00.00000021 2011-02-17 18:33:00+00:00   
121958    2015-01-14 17:35:03.0000007 2015-01-14 17:35:03+00:00   

        pickup_longitude  pickup_latitude  dropoff_longitude  \
353352        -73.993652        40.741543         -73.977974   
360070        -73.993805        40.724579         -73.993805   
372609        -73.959160        40.780750         -73.969116   
550895        -73.952187        40.783951         -73.978645   
444151        -73.977112        40.746834         -73.991104   
...                  ...              ...                ...   
110268        -73.987152        40.750633         -73.979073   
259178        -73.972656        40.764042         -74.013176   
365838        -73.991982        40.749767         -73.989845   
131932        -73.969055        40.761398         -73.990814   
121958        -73.983040        40.742142         -74.002510   

        dropoff_latitude  passenger_count  
353352         40.742352                4  
360070         40.724579                1  
372609         40.761230                1  
550895         40.772602                1  
444151         40.750404                2  
...                  ...              ...  
110268         40.763168                1  
259178         40.707840                2  
365838         40.720551                3  
131932         40.751328                1  
121958         40.727650                1  

[441960 rows x 7 columns]

In [ ]:
train_target

353352     6.0
360070     3.7
372609    10.0
550895     8.9
444151     7.3
          ... 
110268     9.3
259178    18.5
365838    10.1
131932    10.9
121958     9.5
Name: fare_amount, Length: 441960, dtype: float32

In [ ]:
test_inputs = test_df[input_cols]

# Train Hardcoded & Baseline Models

tip - always create  a simple hardcoded odel to establish the mininmum scre that any proper ML model should beat

* Hardcoded Model : always predict ave fare
* Baseline model : Linear Regression

for evaluation use RMSE

In [ ]:
import numpy as np

In [ ]:

class MeanRegressor :
  def fit(self, input, targets):
    self.mean = targets.mean()

  def predict(self, inputs):
    return np.full(inputs.shape[0], self.mean)

In [ ]:
mean_model = MeanRegressor()
mean_model.fit(train_inputs, train_target)
mean_model.mean

np.float32(11.354714)

In [ ]:
train_pred = mean_model.predict(train_inputs)

In [ ]:
train_pred

array([11.354714, 11.354714, 11.354714, ..., 11.354714, 11.354714,
       11.354714], dtype=float32)

In [ ]:
val_pred = mean_model.predict(val_inputs)

In [ ]:
val_pred

array([11.354714, 11.354714, 11.354714, ..., 11.354714, 11.354714,
       11.354714], dtype=float32)

In [ ]:
from sklearn.metrics import mean_squared_error

In [ ]:

def rmse(targets,preds):
  return np.sqrt(mean_squared_error(targets,preds))

In [ ]:
train_rmse = rmse(train_target, train_pred)
train_rmse

np.float64(9.789781840838485)

In [ ]:
val_rmse = rmse(val_target, val_pred)
val_rmse

np.float64(9.89995425435296)

# Train and evaluate a very Base Model



the rmse of linear reg - 9.78 and dumb model i s 9.8 so , linear reg is not much better than your initial dumb model

Make Predictions and Submit to Kaggle -- the best way to improve your model is to try and beet your previous score

In [ ]:
def predict_and_submit(model, test_inputs, fname):
  test_preds = model.predict(test_inputs)
  sub_df = pd.read_csv(data_dir + '/sample_submission.csv')
  sub_df['fare_amount'] = test_preds
  sub_df.to_csv(fname, index=False)
  return sub_df

#Feature Engineering

Tip - Add some features, train the model, evaluate it, keep it they help or drop them, then repeat

* extract the part of dates
* remove the outliers & invalied data
* add distance between drop and pickup
* distance between landmarks

add each feature individally and see the effects of each feature

Extract part of dates
* year
* month
* weekend
* day
* hour

In [ ]:
def add_dateparts(df,col):
  df[col + 'year'] = df[col].dt.year
  df[col + 'month'] = df[col].dt.month
  df[col + 'day'] = df[col].dt.day
  df[col + 'weekday'] = df[col].dt.weekday
  df[col + 'hour'] = df[col].dt.hour

In [ ]:
add_dateparts(train_df,'pickup_datetime')
add_dateparts(val_df,'pickup_datetime')
add_dateparts(test_df,'pickup_datetime')

In [ ]:
train_df

key  fare_amount           pickup_datetime  \
353352    2015-04-12 03:40:38.0000004          6.0 2015-04-12 03:40:38+00:00   
360070  2011-01-26 19:21:00.000000153          3.7 2011-01-26 19:21:00+00:00   
372609    2012-10-03 10:40:17.0000003         10.0 2012-10-03 10:40:17+00:00   
550895    2012-03-14 13:44:27.0000004          8.9 2012-03-14 13:44:27+00:00   
444151   2012-02-05 15:33:00.00000084          7.3 2012-02-05 15:33:00+00:00   
...                               ...          ...                       ...   
110268   2009-09-06 16:12:00.00000051          9.3 2009-09-06 16:12:00+00:00   
259178    2009-04-12 09:58:56.0000001         18.5 2009-04-12 09:58:56+00:00   
365838  2012-07-12 19:30:00.000000266         10.1 2012-07-12 19:30:00+00:00   
131932   2011-02-17 18:33:00.00000021         10.9 2011-02-17 18:33:00+00:00   
121958    2015-01-14 17:35:03.0000007          9.5 2015-01-14 17:35:03+00:00   

        pickup_longitude  pickup_latitude  dropoff_longitude  \
353352        -73.993652        40.741543         -73.977974   
360070        -73.993805        40.724579         -73.993805   
372609        -73.959160        40.780750         -73.969116   
550895        -73.952187        40.783951         -73.978645   
444151        -73.977112        40.746834         -73.991104   
...                  ...              ...                ...   
110268        -73.987152        40.750633         -73.979073   
259178        -73.972656        40.764042         -74.013176   
365838        -73.991982        40.749767         -73.989845   
131932        -73.969055        40.761398         -73.990814   
121958        -73.983040        40.742142         -74.002510   

        dropoff_latitude  passenger_count  pickup_datetimeyear  \
353352         40.742352                4                 2015   
360070         40.724579                1                 2011   
372609         40.761230                1                 2012   
550895         40.772602                1                 2012   
444151         40.750404                2                 2012   
...                  ...              ...                  ...   
110268         40.763168                1                 2009   
259178         40.707840                2                 2009   
365838         40.720551                3                 2012   
131932         40.751328                1                 2011   
121958         40.727650                1                 2015   

        pickup_datetimemonth  pickup_datetimeday  pickup_datetimeweekday  \
353352                     4                  12                       6   
360070                     1                  26                       2   
372609                    10                   3                       2   
550895                     3                  14                       2   
444151                     2                   5                       6   
...                      ...                 ...                     ...   
110268                     9                   6                       6   
259178                     4                  12                       6   
365838                     7                  12                       3   
131932                     2                  17                       3   
121958                     1                  14                       2   

        pickup_datetimehour  
353352                    3  
360070                   19  
372609                   10  
550895                   13  
444151                   15  
...                     ...  
110268                   16  
259178                    9  
365838                   19  
131932                   18  
121958                   17  

[441960 rows x 13 columns]

#Distance between two coordinates

In [ ]:
import math

In [ ]:
def haversine_np(lon1, lat1, lon2, lat2):
    """
    Calculate the great circle distance between two points
    on the earth (specified in decimal degrees)

    All args must be of equal length.
    """
    lon1, lat1, lon2, lat2 = map(np.radians, [lon1, lat1, lon2, lat2])

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = np.sin(dlat / 2.0) ** 2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2.0) ** 2
    c = 2 * np.arcsin(np.sqrt(a))

    km = 6367 * c
    return km


In [ ]:
def add_trip_distance(df):
  df['trip_distance'] = haversine_np(df['pickup_latitude'],
                                  df['pickup_longitude'],
                                  df['dropoff_latitude'],
                                  df['dropoff_longitude'])

In [ ]:
add_trip_distance(train_df)
add_trip_distance(val_df)
add_trip_distance(test_df)

# Add distance from popular landmarks

TIP #11: Creative feature engineering (generally involving human insight or external data) is a lot more effective than excessive hyperparameter tuning. Just one or two good features improve the model's performance drastically.

JFK Airport

LGA Airport

EWR Airport

Times Square

Met Museum

World Trade Center

We'll add the distance from drop location

In [ ]:
jfk_lonlat = -73.7781, 40.6413
lga_lonlat = -73.8740, 40.7769
ewr_lonlat = -74.1745, 40.6895
met_lonlat = -74.0479, 40.7794
wtc_lonlat = -74.0099, 40.7126